### PR TITLE
[FEAT] Attribute Encoder (next version 8.0.0)

### LINKED ISSUE #423: Binary data in model causes JsonEncodingException
| Q                 | A
| ----------------- | ---
| Bug?              | yes, at least for my usecase
| New Feature?      | maybe?
| Framework         | Laravel
| Framework version | 5.6.22
| Package version   | 7.0.0
| PHP version       | 7.2.5

#### Actual Behaviour

Auditing a model that holds binary data results in a `JsonEncodingException`.

#### Expected Behaviour

The audit should convert the binary data before casting it to json.

#### Steps to Reproduce

I have a `MyModel` that has some attribute `foo` wich is stored in binary form in the database. The model comes with `getFooAttribute` and `setFooAttribute` methods that convert from binary to string and back.

When creating a `MyModel`, then the following exception occurs:

```
Illuminate\\Database\\Eloquent\\JsonEncodingException(code: 0): Unable to encode attribute [new_values] for model [App\\Models\\Audit] to JSON: Malformed UTF-8 characters, possibly incorrectly encoded. at PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/JsonEncodingException.php:33)
[stacktrace]
#0 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php(655): Illuminate\\Database\\Eloquent\\JsonEncodingException::forAttribute(Object(App\\Models\\Audit), 'new_values', 'Malformed UTF-8...')
#1 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php(559): Illuminate\\Database\\Eloquent\\Model->castAttributeAsJson('new_values', false)
#2 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(231): Illuminate\\Database\\Eloquent\\Model->setAttribute('new_values', Array)
#3 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(153): Illuminate\\Database\\Eloquent\\Model->fill(Array)
#4 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(294): Illuminate\\Database\\Eloquent\\Model->__construct(Array)
#5 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1068): Illuminate\\Database\\Eloquent\\Model->newInstance(Array)
#6 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(756): Illuminate\\Database\\Eloquent\\Builder->newModelInstance(Array)
#7 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1504): Illuminate\\Database\\Eloquent\\Builder->create(Array)
#8 PATH/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1516): Illuminate\\Database\\Eloquent\\Model->__call('create', Array)
#9 [internal function]: Illuminate\\Database\\Eloquent\\Model::__callStatic('create', Array)
#10 PATH/vendor/owen-it/laravel-auditing/src/Drivers/Database.php(31): call_user_func(Array, Array)
#11 PATH/vendor/owen-it/laravel-auditing/src/Auditor.php(81): OwenIt\\Auditing\\Drivers\\Database->audit(Object(App\\Models\\MyModel))
#12 PATH/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(221): OwenIt\\Auditing\\Auditor->execute(Object(App\\Models\\MyModel))
#13 PATH/vendor/owen-it/laravel-auditing/src/AuditableObserver.php(50): Illuminate\\Support\\Facades\\Facade::__callStatic('execute', Array)
#14 [internal function]: OwenIt\\Auditing\\AuditableObserver->created(Object(App\\Models\\MyModel))
```

#### Possible Solutions

The Auditable currently reads attributes directly from `$this->attributes`, bypassing any accessor ([code](https://github.com/owen-it/laravel-auditing/blob/v7.0.0/src/Auditable.php#L142-L146)).
This is desirable in many cases, where the accessor just beautifies the attribute.

But in cases where the stored data is binary, we need to do some adjustments. This can be
- to use the accessor, assuming that it provides a valid utf-8 string or
- to just base64 encode the data.

How about adding a new property `$auditBinary = ['foo']` to `MyModel` that instructs to base64 encode the `foo` attribute for auditing? The `Audit::getFormattedValue()` function should then decode the data again.

#### Workaround (if you don't want to fix it)

Use `transformAudit` to sanitize `old_values` and `new_values`. But this cannot be easily integrated in `Audit::getModified()`.

### PR DESCRIPTION
**Pull request by quetzyg**
This pull request introduces a new feature, allowing to store binary data (encoded in Base64), which should prevent a `JsonEncodingException` when auditing an Eloquent model with binary data attributes (fixes #423).

Work done:
- [X] Implemented new `AttributeModifier` interfaces;
- [X] Implemented `Base64Encoder` class;
- [X] Simplified configuration;
- [X] Updated/added tests;
- [X] Minor code cleanups;

Given the changes that had to be made, this new feature breaks backwards compatibility with the previous major version (mainly with the `AuditRedactor` feature).

### CODE REVIEW

File: src/Contracts/AttributeEncoder.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public static function encode($value): string;
+
+    /**
+     * Decode an attribute value.
+     *
+     * @param mixed $value
+     *
+     * @return string
+     */
+    public static function decode($value): string;
```
Comment (SebastianS90) [inline src/Contracts/AttributeEncoder.php]: @quetzyg Any reason why the encoded/decoded value must be a `string`? A custom encoder could also return `null` here.

My use case (the auditable model has binary IP address attributes that may be null):
```php
    public static function encode($value)
    {
        return is_null($value) ? null : inet_ntop($value);
    }

    public static function decode($value)
    {
        return $value ? inet_pton($value) : null;
    }
```

Comment (quetzyg) [reply src/Contracts/AttributeEncoder.php]: Good point! I'll update both methods so that there's no return type enforced.

Comment (quetzyg) [reply src/Contracts/AttributeEncoder.php]: [Updated](https://github.com/owen-it/laravel-auditing/commit/277f1c1f151ba7a77d5e967098f4fe2d005dfa71)

**Comment by RRStoyanov** [comment]
Hello, when do you plan to release 8.0.0 version? It's in the changelog but new feature isn't added yet to master? 🤔

**Comment by quetzyg** [comment]
What do you mean? It's right [there](https://github.com/owen-it/laravel-auditing/tree/v8.0.0).

**Pull request merged**